### PR TITLE
Remove the BMCClient interface, ginkgo, and gomega

### PR DESCRIPTION
**Pull request by jacobweinstock**
## Why is this needed
The BMCClient interface was unnecessary and added complexity to testing. bmclib already has the capabilities to specify mock providers. Removing this interface also allowed us to remove ginkgo and gomega. These frameworks were unneeded and overly complex. This PR simplifies tests, increases coverage, and sets us up to build out more tests quickly and simply.

Fixes: #59 

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: controllers/job_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	tests := map[string]struct {
+		machine   *v1alpha1.Machine
+		secret    *corev1.Secret
+		job       *v1alpha1.Job
+		shouldErr bool
+		testAll   bool
+	}{
+		"success taskless job":    {machine: createMachine(), secret: createSecret(), job: createJob("test", createMachine())},
+		"failure unknown machine": {machine: &v1alpha1.Machine{}, secret: createSecret(), job: createJob("test", createMachine()), shouldErr: true},
+		"success power on job":    {machine: createMachine(), secret: createSecret(), job: createJob("test", createMachine(), getAction("PowerOn")), testAll: true},
```
Comment (chrisdoherty4) [inline controllers/job_test.go:26]: Mind splitting the object construction over lines so we don't run off the page?

**Review by chrisdoherty4 — COMMENTED** [review]
Some great improvements. Glad to be removing Ginkgo.

File: controllers/task_controller_test.go [hunk 2]
```diff
@@ -6,193 +6,59 @@ import (
...
+	default:
+		return v1alpha1.Action{}
 	}
 }
 
-func TestTaskReconciler_ReconcileBootDevice(t *testing.T) {
-	for name, tt := range map[string]struct {
-		configureClientCalls func(expect *mocks.MockBMCClientMockRecorder)
-		action               v1alpha1.Action
+func TestTaskReconcile(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply controllers/task_controller_test.go]: this was mostly a translation of the ginko/gomega tests. The idea was to start by removing those libs and then refactoring the tests. I can do the refactor in this PR if you'd like.

File: api/v1alpha1/job.go [hunk 3]
```diff
@@ -144,9 +144,9 @@ func FormatTaskName(job Job, n int) string {
 //+kubebuilder:subresource:status
 //+kubebuilder:resource:path=jobs,scope=Namespaced,categories=tinkerbell,singular=job,shortName=j
 
-// Job is the Schema for the bmcjobs API
+// Job is the Schema for the bmcjobs API.
 type Job struct {
-	metav1.TypeMeta   `json:",inline"`
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply api/v1alpha1/job.go:149]: `inline` is not a recognized struct tag in Go. I believe the only reason to keep them is if we have any code generation that depends on it. 

ref: 
https://github.com/golang/go/issues/6213
https://github.com/kubernetes/kubernetes/pull/117206

File: controllers/job.go [hunk 4]
```diff
@@ -77,41 +76,41 @@ func (r *JobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
...
 	// Create a patch from the initial Job object
 	// Patch is used to update Status after reconciliation
 	jobPatch := client.MergeFrom(job.DeepCopy())
 
-	return r.reconcile(ctx, job, jobPatch, logger)
+	return r.doReconcile(ctx, job, jobPatch)
 }
 
-func (r *JobReconciler) reconcile(ctx context.Context, job *bmcv1alpha1.Job, jobPatch client.Patch, logger logr.Logger) (ctrl.Result, error) {
+func (r *JobReconciler) doReconcile(ctx context.Context, job *v1alpha1.Job, jobPatch client.Patch) (ctrl.Result, error) {
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply controllers/job.go:91]: this was a linting issue that came up. I'm open to changing this. I can revert back and add a `notlint` or change the name entirely. Thoughts?

File: controllers/job_test.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func TestJobReconcile(t *testing.T) {
+	tests := map[string]struct {
+		machine   *v1alpha1.Machine
+		secret    *corev1.Secret
+		job       *v1alpha1.Job
+		shouldErr bool
+		testAll   bool
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply controllers/job_test.go:22]: `testAll` allows this test to do more testing than just if `reconciler.Reconcile` was successful. This is an artifact of translating the ginkgo/gomega tests. I will refactor, just depends on if you would like that done in this PR or a follow-up. This PR is fairly large as is.

File: controllers/job_test.go (hunk 1 above)
Comment (jacobweinstock) [reply controllers/job_test.go:26]: sure thing.

File: api/v1alpha1/job.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisdoherty4) [reply api/v1alpha1/job.go:149]: Thanks, this has plagued me with the linting and I could never find anything useful about it.

**Review by chrisdoherty4 — APPROVED** [review]

**Comment by chrisdoherty4** [comment]
@mergifyio queue

**Comment by jacobweinstock** [comment]
@mergifyio queue

**Pull request merged**